File: frontend/js/views/split_shift_premium.js
// Split-shift premium generator — one hour at minimum wage, offset by earnings
// above minimum wage, via /calc/split-shift-premium.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';

const esc = (s) => String(s).replace(/[&<>]/g, (c) => ({ '&': '&amp;', '<': '&lt;', '>': '&gt;' }[c]));
const money = (n) => (n == null ? '—' : '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 }));
let LAST_DOC = null;

export async function renderSplitShiftPremium(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.splitshift.h1.title">// SPLIT-SHIFT PREMIUM</span></h1>
        <p class="muted small" data-i18n="view.splitshift.hint.intro">
            When an employer schedules a work day with an unpaid, non-meal gap (a split shift), California law
            owes the employee one additional hour of pay at the minimum wage — offset by the amount the
            employee's earnings for the day already exceed the minimum wage. It computes the one-hour premium,
            the offset, and the net premium owed. Drafting aid, not legal advice.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.splitshift.h2.inputs">Premium inputs</h2>
            <form id="split-form" class="inline-form">
                <label><span data-i18n="view.splitshift.label.state">State</span>
                    <input type="text" name="state" value="California" required></label>
                <label><span data-i18n="view.splitshift.label.employer">Employer</span>
                    <input type="text" name="employer_name" value=""></label>
                <label><span data-i18n="view.splitshift.label.employee">Employee</span>
                    <input type="text" name="employee_name" value=""></label>
                <label><span data-i18n="view.splitshift.label.minwage">Minimum wage ($/hr)</span>
                    <input type="number" step="0.25" min="0" name="min_wage_usd" value="16" required></label>
                <label><span data-i18n="view.splitshift.label.rate">Regular rate ($/hr)</span>
                    <input type="number" step="0.01" min="0" name="regular_rate_usd" value="16" required></label>
                <label><span data-i18n="view.splitshift.label.hours">Hours worked</span>
                    <input type="number" step="0.25" min="0" name="hours_worked" value="8" required></label>
                <label><span data-i18n="view.splitshift.label.shiftdate">Shift date</span>
                    <input type="date" name="shift_date" value="2026-06-20" required></label>
                <label><span data-i18n="view.splitshift.label.date">Statement date</span>
                    <input type="date" name="date" value="2026-07-01" required></label>
                <label><span data-i18n="view.splitshift.label.statute">Statute citation (optional)</span>
                    <input type="text" name="statute_citation" value="" placeholder="${esc(t('view.splitshift.ph.statute'))}"></label>
            </form>
        </div>
        <div id="split-result" class="chart-panel lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#split-form');
    const generate = async () => {
        const fd = new FormData(form);
        const body = {
            employer_name: (fd.get('employer_name') || '').trim(),
            employee_name: (fd.get('employee_name') || '').trim(),
            min_wage_usd: Number(fd.get('min_wage_usd')) || 0,
            regular_rate_usd: Number(fd.get('regular_rate_usd')) || 0,
            hours_worked: Number(fd.get('hours_worked')) || 0,
            shift_date: fd.get('shift_date'),
            date: fd.get('date'),
            state: (fd.get('state') || '').trim(),
            statute_citation: (fd.get('statute_citation') || '').trim(),
        };
        try {
            const doc = await api.calcSplitShiftPremium(body);
            if (!viewIsCurrent(tok)) return;
            renderResult(mount, doc);
        } catch (err) {
            showToast(err.message || t('view.splitshift.toast.error'), { level: 'error' });
        }
    };
    const live = debounce(generate, 250);
    form.addEventListener('input', () => { live(); });
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function docToText(doc) {
    const lines = [doc.title.toUpperCase()];
    if (doc.statutory_citation) lines.push(doc.statutory_citation);
    lines.push('');
    for (const c of doc.clauses) lines.push(c.heading, c.body, '');
    return lines.join('\n');
}

function renderResult(mount, doc) {
    LAST_DOC = doc;
    const offsetNote = doc.fully_offset
        ? ` <span class="muted small" data-i18n="view.splitshift.offset">(fully offset)</span>`
        : '';
    const el = mount.querySelector('#split-result');
    el.innerHTML = `
        <div class="lpv-bar">
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.splitshift.card.net">Net premium</div>
                    <div class="value">${money(doc.net_premium_usd)}${offsetNote}</div></div>
                <div class="card"><div class="label" data-i18n="view.splitshift.card.onehour">One-hour premium</div>
                    <div class="value">${money(doc.one_hour_premium_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.splitshift.card.offset">Earnings above min</div>
                    <div class="value">${money(doc.earnings_above_min_usd)}</div></div>
            </div>
            <div class="btn-row-inline">
                <button class="btn btn-secondary" id="split-copy" type="button" data-i18n="view.splitshift.btn.copy">Copy</button>
                <button class="btn btn-secondary" id="split-download" type="button" data-i18n="view.splitshift.btn.download">Download .txt</button>
            </div>
        </div>
        <pre class="small">${esc(docToText(doc))}</pre>
    `;
    applyUiI18n(el);

    el.querySelector('#split-copy').addEventListener('click', async () => {
        try {
            await navigator.clipboard.writeText(docToText(LAST_DOC));
            showToast(t('view.splitshift.toast.copied'), { level: 'success' });
        } catch (e) {
            showToast(t('view.splitshift.toast.copy_failed', { err: e.message || String(e) }), { level: 'error' });
        }
    });
    el.querySelector('#split-download').addEventListener('click', () => {
        const blob = new Blob([docToText(LAST_DOC)], { type: 'text/plain' });
        const url = URL.createObjectURL(blob);
        const a = document.createElement('a');
        a.href = url;
        a.download = 'split-shift-premium.txt';
        a.click();
        URL.revokeObjectURL(url);
    });
}

File: crates/traderview-core/src/split_shift_premium.rs
//! Split-shift premium — when an employer schedules a work day with an unpaid,
//! non-meal gap (a "split shift"), California law owes the employee one additional
//! hour of pay at the minimum wage. That premium is offset by the amount by which
//! the employee's earnings for the day already exceed the minimum wage: premium =
//! max(minimum wage − (regular rate − minimum wage) × hours worked, 0). This
//! computes the one-hour premium, the offset, and the net premium owed. No existing
//! generator computes this minimum-wage offset. Drafting aid, not legal advice.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Deserialize)]
pub struct SplitShiftInput {
    pub employer_name: String,
    pub employee_name: String,
    /// Applicable minimum wage, per hour.
    pub min_wage_usd: f64,
    /// Employee's regular rate of pay, per hour.
    pub regular_rate_usd: f64,
    /// Hours actually worked on the split-shift day.
    pub hours_worked: f64,
    pub shift_date: String,
    pub date: String,
    pub state: String,
    #[serde(default)]
    pub statute_citation: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct DocClause {
    pub heading: String,
    pub body: String,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct SplitShiftPremium {
    pub title: String,
    /// One hour of pay at the minimum wage.
    pub one_hour_premium_usd: f64,
    /// Earnings above the minimum wage for the hours worked.
    pub earnings_above_min_usd: f64,
    /// One-hour premium less the offset, floored at 0.
    pub net_premium_usd: f64,
    /// True when the offset fully eliminated the premium.
    pub fully_offset: bool,
    pub statutory_citation: String,
    pub clauses: Vec<DocClause>,
}

fn cents(x: f64) -> f64 {
    (x * 100.0).round() / 100.0
}

fn money(v: f64) -> String {
    format!("${:.2}", v)
}

pub fn generate(i: &SplitShiftInput) -> SplitShiftPremium {
    let one_hour = cents(i.min_wage_usd);
    let above_min = cents((i.regular_rate_usd - i.min_wage_usd).max(0.0) * i.hours_worked);
    let net = cents((one_hour - above_min).max(0.0));
    let fully_offset = net <= 0.0;

    let calc_body = format!(
        "The split-shift premium is one hour at the minimum wage of {} = {}. The employee's earnings above the minimum wage are {} (({} − {}) × {} hours), which offsets the premium. The net split-shift premium owed is {}.",
        money(i.min_wage_usd),
        money(one_hour),
        money(above_min),
        money(i.regular_rate_usd),
        money(i.min_wage_usd),
        i.hours_worked,
        money(net)
    );

    let citation = i.statute_citation.trim();
    let governing = if citation.is_empty() {
        format!("The split-shift premium is owed under the wage laws of the State of {}.", i.state)
    } else {
        format!("The split-shift premium is owed under the wage laws of the State of {} ({}).", i.state, citation)
    };

    let clauses = vec![
        DocClause {
            heading: "Parties".into(),
            body: format!(
                "Employer: {}\nEmployee: {}\nShift date: {}\nStatement date: {}",
                i.employer_name, i.employee_name, i.shift_date, i.date
            ),
        },
        DocClause {
            heading: "1. Split Shift".into(),
            body: "A split shift is a schedule interrupted by an unpaid, non-meal period. For each split-shift day, the employee is owed one additional hour of pay at the minimum wage, reduced by earnings already above the minimum wage that day.".into(),
        },
        DocClause { heading: "2. Calculation".into(), body: calc_body },
        DocClause {
            heading: "3. Payment".into(),
            body: if fully_offset {
                "No split-shift premium is owed: the employee's earnings above the minimum wage fully offset the one-hour premium.".to_string()
            } else {
                format!("The employer owes the employee a split-shift premium of {} for the shift.", money(net))
            },
        },
        DocClause { heading: "4. Governing Law".into(), body: governing },
        DocClause {
            heading: "Certification".into(),
            body: format!(
                "Employer: ____________________  Date: __________\n{}",
                i.employer_name
            ),
        },
    ];

    SplitShiftPremium {
        title: "Split-Shift Premium Statement".into(),
        one_hour_premium_usd: one_hour,
        earnings_above_min_usd: above_min,
        net_premium_usd: net,
        fully_offset,
        statutory_citation: citation.to_string(),
        clauses,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 0.01
    }

    fn base() -> SplitShiftInput {
        SplitShiftInput {
            employer_name: "Acme Co".into(),
            employee_name: "Pat Worker".into(),
            min_wage_usd: 16.0,
            regular_rate_usd: 16.0,
            hours_worked: 8.0,
            shift_date: "2026-06-20".into(),
            date: "2026-07-01".into(),
            state: "California".into(),
            statute_citation: String::new(),
        }
    }

    #[test]
    fn full_premium_at_minimum_wage() {
        let d = generate(&base());
        assert!(close(d.one_hour_premium_usd, 16.0));
        assert!(close(d.earnings_above_min_usd, 0.0));
        assert!(close(d.net_premium_usd, 16.0));
        assert!(!d.fully_offset);
    }

    #[test]
    fn fully_offset_above_minimum() {
        let d = generate(&SplitShiftInput { regular_rate_usd: 18.0, ..base() });
        // (18−16) × 8 = 16 ≥ 16 → offset fully.
        assert!(close(d.earnings_above_min_usd, 16.0));
        assert!(close(d.net_premium_usd, 0.0));
        assert!(d.fully_offset);
    }

    #[test]
    fn partial_offset() {
        let d = generate(&SplitShiftInput { regular_rate_usd: 17.0, ..base() });
        // (17−16) × 8 = 8 → premium 16 − 8 = 8.
        assert!(close(d.earnings_above_min_usd, 8.0));
        assert!(close(d.net_premium_usd, 8.0));
        assert!(!d.fully_offset);
    }

    #[test]
    fn fewer_hours_less_offset() {
        let d = generate(&SplitShiftInput { regular_rate_usd: 20.0, hours_worked: 4.0, ..base() });
        // (20−16) × 4 = 16 → fully offset.
        assert!(close(d.net_premium_usd, 0.0));
    }

    #[test]
    fn statute_citation_echoed() {
        let d = generate(&SplitShiftInput { statute_citation: "Cal. Code Regs. tit. 8 § 11040".into(), ..base() });
        assert_eq!(d.statutory_citation, "Cal. Code Regs. tit. 8 § 11040");
        assert!(d.clauses.iter().any(|c| c.body.contains("Cal. Code Regs. tit. 8 § 11040")));
    }
}
